File: app/src/main/java/com/dynamsoft/sample/dbrcamerapreview/util/CameraUtil.java
package com.dynamsoft.sample.dbrcamerapreview.util;

import android.content.Context;
import android.graphics.Point;
import android.graphics.PointF;
import android.graphics.Rect;
import android.graphics.RectF;
import android.util.Size;
import android.view.Display;
import android.view.Surface;
import android.view.WindowManager;

public class CameraUtil {
    /**
     * Clamps x to between min and max (inclusive on both ends, x = min --> min,
     * x = max --> max).
     */
    public static int clamp(int x, int min, int max) {
        if (x > max) {
            return max;
        }
        if (x < min) {
            return min;
        }
        return x;
    }

    /**
     * Clamps x to between min and max (inclusive on both ends, x = min --> min,
     * x = max --> max).
     */
    public static float clamp(float x, float min, float max) {
        if (x > max) {
            return max;
        }
        if (x < min) {
            return min;
        }
        return x;
    }

    public static void inlineRectToRectF(RectF rectF, Rect rect) {
        rect.left = Math.round(rectF.left);
        rect.top = Math.round(rectF.top);
        rect.right = Math.round(rectF.right);
        rect.bottom = Math.round(rectF.bottom);
    }

    public static Rect rectFToRect(RectF rectF) {
        Rect rect = new Rect();
        inlineRectToRectF(rectF, rect);
        return rect;
    }

    public static RectF rectToRectF(Rect r) {
        return new RectF(r.left, r.top, r.right, r.bottom);
    }

    /**
     * Linear interpolation between a and b by the fraction t. t = 0 --> a, t =
     * 1 --> b.
     */
    public static float lerp(float a, float b, float t) {
        return a + t * (b - a);
    }

    /**
     * Given (nx, ny) \in [0, 1]^2, in the display's portrait coordinate system,
     * returns normalized sensor coordinates \in [0, 1]^2 depending on how the
     * sensor's orientation \in {0, 90, 180, 270}.
     * <p>
     * Returns null if sensorOrientation is not one of the above.
     * </p>
     */
    public static PointF normalizedSensorCoordsForNormalizedDisplayCoords(
            float nx, float ny, int sensorOrientation) {
        switch (sensorOrientation) {
            case 0:
                return new PointF(nx, ny);
            case 90:
                return new PointF(ny, 1.0f - nx);
            case 180:
                return new PointF(1.0f - nx, 1.0f - ny);
            case 270:
                return new PointF(1.0f - ny, nx);
            default:
                return null;
        }
    }

    public static Rect boundaryRotate(Point orgPt, Rect rect, boolean bLeft) {
        float orgx = orgPt.x;
        float orgy = orgPt.y;

        float rotatex = orgy;
        float rotatey = orgx;
        float[] currentBoundary = new float[8];
        currentBoundary[0] = rect.left;
        currentBoundary[1] = rect.top;

        currentBoundary[2] = rect.right;
        currentBoundary[3] = rect.top;

        currentBoundary[4] = rect.right;
        currentBoundary[5] = rect.bottom;

        currentBoundary[6] = rect.left;
        currentBoundary[7] = rect.bottom;

        int[] rotateBoundary = new int[8];
        if (bLeft) {
            rotateBoundary[6] = (int) ((currentBoundary[0] - orgx) * Math.cos(Math.PI * 0.5f) + (currentBoundary[1] - orgy) * Math.sin(Math.PI * 0.5f) + rotatex);
            rotateBoundary[7] = (int) (-(currentBoundary[0] - orgx) * Math.sin(Math.PI * 0.5f) + (currentBoundary[1] - orgy) * Math.cos(Math.PI * 0.5f) + rotatey);

            rotateBoundary[0] = (int) ((currentBoundary[2] - orgx) * Math.cos(Math.PI * 0.5f) + (currentBoundary[3] - orgy) * Math.sin(Math.PI * 0.5f) + rotatex);
            rotateBoundary[1] = (int) (-(currentBoundary[2] - orgx) * Math.sin(Math.PI * 0.5f) + (currentBoundary[3] - orgy) * Math.cos(Math.PI * 0.5f) + rotatey);

            rotateBoundary[2] = (int) ((currentBoundary[4] - orgx) * Math.cos(Math.PI * 0.5f) + (currentBoundary[5] - orgy) * Math.sin(Math.PI * 0.5f) + rotatex);
            rotateBoundary[3] = (int) (-(currentBoundary[4] - orgx) * Math.sin(Math.PI * 0.5f) + (currentBoundary[5] - orgy) * Math.cos(Math.PI * 0.5f) + rotatey);

            rotateBoundary[4] = (int) ((currentBoundary[6] - orgx) * Math.cos(Math.PI * 0.5f) + (currentBoundary[7] - orgy) * Math.sin(Math.PI * 0.5f) + rotatex);
            rotateBoundary[5] = (int) (-(currentBoundary[6] - orgx) * Math.sin(Math.PI * 0.5f) + (currentBoundary[7] - orgy) * Math.cos(Math.PI * 0.5f) + rotatey);
        } else {
            rotateBoundary[2] = (int) ((currentBoundary[0] - orgx) * Math.cos(Math.PI * 0.5f) - (currentBoundary[1] - orgy) * Math.sin(Math.PI * 0.5f) + rotatex);
            rotateBoundary[3] = (int) (((currentBoundary[0] - orgx) * Math.sin(Math.PI * 0.5f)) + ((currentBoundary[1] - orgy) * Math.cos(Math.PI * 0.5f)) + rotatey);

            rotateBoundary[4] = (int) ((currentBoundary[2] - orgx) * Math.cos(Math.PI * 0.5f) - (currentBoundary[3] - orgy) * Math.sin(Math.PI * 0.5f) + rotatex);
            rotateBoundary[5] = (int) ((currentBoundary[2] - orgx) * Math.sin(Math.PI * 0.5f) + (currentBoundary[3] - orgy) * Math.cos(Math.PI * 0.5f) + rotatey);

            rotateBoundary[6] = (int) ((currentBoundary[4] - orgx) * Math.cos(Math.PI * 0.5f) - (currentBoundary[5] - orgy) * Math.sin(Math.PI * 0.5f) + rotatex);
            rotateBoundary[7] = (int) ((currentBoundary[4] - orgx) * Math.sin(Math.PI * 0.5f) + (currentBoundary[5] - orgy) * Math.cos(Math.PI * 0.5f) + rotatey);

            rotateBoundary[0] = (int) ((currentBoundary[6] - orgx) * Math.cos(Math.PI * 0.5f) - (currentBoundary[7] - orgy) * Math.sin(Math.PI * 0.5f) + rotatex);
            rotateBoundary[1] = (int) ((currentBoundary[6] - orgx) * Math.sin(Math.PI * 0.5f) + (currentBoundary[7] - orgy) * Math.cos(Math.PI * 0.5f) + rotatey);
        }

        Rect rotateRect = new Rect(rotateBoundary[0], rotateBoundary[1], rotateBoundary[2], rotateBoundary[5]);

        return rotateRect;
    }

    public static Rect boundaryRotate180(Point orgPt, Rect rect) {
        float orgx = orgPt.x;
        float orgy = orgPt.y;

        float rotatex = orgy;
        float rotatey = orgx;
        float[] currentBoundary = new float[8];
        currentBoundary[0] = rect.left;
        currentBoundary[1] = rect.top;

        currentBoundary[2] = rect.right;
        currentBoundary[3] = rect.top;

        currentBoundary[4] = rect.right;
        currentBoundary[5] = rect.bottom;

        currentBoundary[6] = rect.left;
        currentBoundary[7] = rect.bottom;
        int[] rotateBoundary = new int[8];
        rotateBoundary[4] = (int) (orgx - (currentBoundary[0] - orgx));
        rotateBoundary[5] = (int) (orgy - (currentBoundary[1] - orgy));

        rotateBoundary[6] = (int) (orgx - (currentBoundary[2] - orgx));
        rotateBoundary[7] = (int) (orgy - (currentBoundary[3] - orgy));

        rotateBoundary[0] = (int) (orgx - (currentBoundary[4] - orgx));
        rotateBoundary[1] = (int) (orgy - (currentBoundary[5] - orgy));

        rotateBoundary[2] = (int) (orgx - (currentBoundary[6] - orgx));
        rotateBoundary[3] = (int) (orgy - (currentBoundary[7] - orgy));
        Rect rotateRect = new Rect(rotateBoundary[0], rotateBoundary[1], rotateBoundary[2], rotateBoundary[5]);

        return rotateRect;
    }

    public static int getOrientationDisplayOffset(Context context, int nSensorOrientation) {
        int mDisplayOffset = 0;
        Display display = ((WindowManager) context.getSystemService(Context.WINDOW_SERVICE)).getDefaultDisplay();
        switch (display.getRotation()) {
            case Surface.ROTATION_0:
                mDisplayOffset = 0;
                break;
            case Surface.ROTATION_90:
                mDisplayOffset = 90;
                break;
            case Surface.ROTATION_180:
                mDisplayOffset = 180;
                break;
            case Surface.ROTATION_270:
                mDisplayOffset = 270;
                break;
            default:
                mDisplayOffset = 0;
                break;
        }
        //		if (mFacing == Facing.FRONT) {
//			// Here we had ((mSensorOffset - mDisplayOffset) + 360 + 180) % 360
//			// And it seemed to give the same results for various combinations, but not for all (e.g. 0 - 270).
//			return (360 - ((mSensorOffset + mDisplayOffset) % 360)) % 360;
//		} else
        {
            int nOrientationDisplayOffset = (nSensorOrientation - mDisplayOffset + 360) % 360;
            return nOrientationDisplayOffset;
        }
    }

    public static Rect ConvertViewRegionToVideoFrameRegion(Rect viewRegion, Rect frameSize, int nOrientationDisplayOffset, Size szCameraView) {
        Rect convertRegion;
        final int rotateDegree = nOrientationDisplayOffset;
        if (rotateDegree == 90) {
            convertRegion = boundaryRotate(new Point(szCameraView.getWidth() / 2, szCameraView.getHeight() / 2), viewRegion, true);
        } else if (rotateDegree == 180) {
            convertRegion = boundaryRotate180(new Point(szCameraView.getWidth() / 2, szCameraView.getHeight() / 2), viewRegion);
        } else if (nOrientationDisplayOffset == 270) {
            convertRegion = boundaryRotate(new Point(szCameraView.getWidth() / 2, szCameraView.getHeight() / 2), viewRegion, false);
        } else {
            convertRegion = viewRegion;
        }

        int nViewW = szCameraView.getWidth();
        int nViewH = szCameraView.getHeight();
        float fScaleH = (nOrientationDisplayOffset % 180 == 0) ? 1.0f * frameSize.height() / nViewH : 1.0f * frameSize.height() / nViewW;
        float fScaleW = (nOrientationDisplayOffset % 180 == 0) ? 1.0f * frameSize.width() / nViewW : 1.0f * frameSize.width() / nViewH;
        float fScale = (fScaleH > fScaleW) ? fScaleW : fScaleH;
        int boxLeft = (int) (convertRegion.left * fScale);
        int boxTop = (int) (convertRegion.top * fScale);
        int boxWidth = (int) (convertRegion.width() * fScale);
        int boxHeight = (int) (convertRegion.height() * fScale);
        Rect frameRegion = new Rect(boxLeft, boxTop, boxWidth + boxLeft, boxTop + boxHeight);
        return frameRegion;
    }

    public static Rect ConvertFrameRegionToViewRegion(Rect frameRegion, Rect frameSize, int nOrientationDisplayOffset, Size szCameraView) {

        Rect imageRect = frameSize;
        Rect roateRect = frameRegion;
        int rotateDegree = nOrientationDisplayOffset;

        if (rotateDegree == 90) {
            roateRect = boundaryRotate(new Point(imageRect.width() / 2, imageRect.height() / 2), frameRegion, false);
        } else if (rotateDegree == 180) {
            roateRect = boundaryRotate180(new Point(imageRect.width() / 2, imageRect.height() / 2), frameRegion);
        } else if (nOrientationDisplayOffset == 270) {
            roateRect = boundaryRotate(new Point(imageRect.width() / 2, imageRect.height() / 2), frameRegion, true);
        }

        int nViewW = szCameraView.getWidth();
        int nViewH = szCameraView.getHeight();

        float fScaleH = (nOrientationDisplayOffset % 180 == 0) ? 1.0f * nViewH / imageRect.height() : 1.0f * nViewH / imageRect.width();
        float fScaleW = (nOrientationDisplayOffset % 180 == 0) ? 1.0f * nViewW / imageRect.width() : 1.0f * nViewW / imageRect.height();
        float fScale = (fScaleH > fScaleW) ? fScaleW : fScaleH;

        int boxLeft = (int) (roateRect.left * fScale);
        int boxTop = (int) (roateRect.top * fScale);
        int boxWidth = (int) (roateRect.width() * fScale);
        int boxHeight = (int) (roateRect.height() * fScale);
        Rect viewRegion = new Rect(boxLeft, boxTop, boxWidth + boxLeft, boxTop + boxHeight);
        return viewRegion;

    }
}

File: client/app/src/main/java/com/dynamsoft/sample/dbrcamerapreview/util/CameraUtil.kt
package com.dynamsoft.sample.dbrcamerapreview.util

import android.content.Context
import android.graphics.Point
import android.graphics.PointF
import android.graphics.Rect
import android.graphics.RectF
import android.util.Size
import android.view.Surface
import android.view.WindowManager
import java.lang.IllegalArgumentException
import kotlin.math.cos
import kotlin.math.roundToInt
import kotlin.math.sin

object CameraUtil {
    /**
     * Clamps x to between min and max (inclusive on both ends, x = min --> min, x = max --> max).
     */
    fun <T> clamp(x: T, min: T, max: T): T where T: Number, T: Comparable<T> {
        return when {
            x > max -> max
            x < min -> min
            else -> x
        }
    }

    fun inlineRectToRectF(rectF: RectF, rect: Rect) {
        rect.left = rectF.left.roundToInt()
        rect.top = rectF.top.roundToInt()
        rect.right = rectF.right.roundToInt()
        rect.bottom = rectF.bottom.roundToInt()
    }

    fun rectFToRect(rectF: RectF): Rect {
        val rect = Rect()
        inlineRectToRectF(rectF, rect)
        return rect
    }

    fun rectToRectF(r: Rect): RectF {
        return RectF(r.left.toFloat(), r.top.toFloat(), r.right.toFloat(), r.bottom.toFloat())
    }

    /**
     * Linear interpolation between a and b by the fraction t.
     * t = 0 --> a, t = 1 --> b.
     */
    fun linearInterpolation(a: Float, b: Float, t: Float): Float {
        return a + t * (b - a)
    }

    /**
     * Given (nx, ny) \in [0, 1]^2, in the display's portrait coordinate system,
     * returns normalized sensor coordinates \in [0, 1]^2 depending on how the
     * sensor's orientation \in {0, 90, 180, 270}.
     *
     * Returns null if sensorOrientation is not one of the above.
     *
     */
    fun normalizedSensorCoordsForNormalizedDisplayCoords(nx: Float, ny: Float, sensorOrientation: Int): PointF {
        return when (sensorOrientation) {
            0 -> PointF(nx, ny)
            90 -> PointF(ny, 1.0f - nx)
            180 -> PointF(1.0f - nx, 1.0f - ny)
            270 -> PointF(1.0f - ny, nx)
            else -> throw IllegalArgumentException("Invalid sensorOrientation: $sensorOrientation.")
        }
    }

    private fun getBoundary(rect: Rect): FloatArray {
        val boundary = FloatArray(8)
        boundary[0] = rect.left.toFloat()
        boundary[1] = rect.top.toFloat()
        boundary[2] = rect.right.toFloat()
        boundary[3] = rect.top.toFloat()
        boundary[4] = rect.right.toFloat()
        boundary[5] = rect.bottom.toFloat()
        boundary[6] = rect.left.toFloat()
        boundary[7] = rect.bottom.toFloat()
        return boundary
    }

    fun boundaryRotate(orgPt: Point, rect: Rect, bLeft: Boolean): Rect {
        val orgX = orgPt.x.toFloat()
        val orgY = orgPt.y.toFloat()
        val currentBoundary: FloatArray = getBoundary(rect)
        val rotateBoundary = IntArray(8)
        val cosHalfPi = cos(Math.PI * 0.5f)
        val sinHalfPi = sin(Math.PI * 0.5f)

        if (bLeft) {
            rotateBoundary[6] = ((currentBoundary[0] - orgX) * cosHalfPi + (currentBoundary[1] - orgY) * sinHalfPi + orgY).toInt()
            rotateBoundary[7] = (-(currentBoundary[0] - orgX) * sinHalfPi + (currentBoundary[1] - orgY) * cosHalfPi + orgX).toInt()
            rotateBoundary[0] = ((currentBoundary[2] - orgX) * cosHalfPi + (currentBoundary[3] - orgY) * sinHalfPi + orgY).toInt()
            rotateBoundary[1] = (-(currentBoundary[2] - orgX) * sinHalfPi + (currentBoundary[3] - orgY) * cosHalfPi + orgX).toInt()
            rotateBoundary[2] = ((currentBoundary[4] - orgX) * cosHalfPi + (currentBoundary[5] - orgY) * sinHalfPi + orgY).toInt()
            rotateBoundary[3] = (-(currentBoundary[4] - orgX) * sinHalfPi + (currentBoundary[5] - orgY) * cosHalfPi + orgX).toInt()
            rotateBoundary[4] = ((currentBoundary[6] - orgX) * cosHalfPi + (currentBoundary[7] - orgY) * sinHalfPi + orgY).toInt()
            rotateBoundary[5] = (-(currentBoundary[6] - orgX) * sinHalfPi + (currentBoundary[7] - orgY) * cosHalfPi + orgX).toInt()
        } else {
            rotateBoundary[2] = ((currentBoundary[0] - orgX) * cosHalfPi - (currentBoundary[1] - orgY) * sinHalfPi + orgY).toInt()
            rotateBoundary[3] = ((currentBoundary[0] - orgX) * sinHalfPi + (currentBoundary[1] - orgY) * cosHalfPi + orgX).toInt()
            rotateBoundary[4] = ((currentBoundary[2] - orgX) * cosHalfPi - (currentBoundary[3] - orgY) * sinHalfPi + orgY).toInt()
            rotateBoundary[5] = ((currentBoundary[2] - orgX) * sinHalfPi + (currentBoundary[3] - orgY) * cosHalfPi + orgX).toInt()
            rotateBoundary[6] = ((currentBoundary[4] - orgX) * cosHalfPi - (currentBoundary[5] - orgY) * sinHalfPi + orgY).toInt()
            rotateBoundary[7] = ((currentBoundary[4] - orgX) * sinHalfPi + (currentBoundary[5] - orgY) * cosHalfPi + orgX).toInt()
            rotateBoundary[0] = ((currentBoundary[6] - orgX) * cosHalfPi - (currentBoundary[7] - orgY) * sinHalfPi + orgY).toInt()
            rotateBoundary[1] = ((currentBoundary[6] - orgX) * sinHalfPi + (currentBoundary[7] - orgY) * cosHalfPi + orgX).toInt()
        }

        return Rect(rotateBoundary[0], rotateBoundary[1], rotateBoundary[2], rotateBoundary[5])
    }

    fun boundaryRotate180(orgPt: Point, rect: Rect): Rect {
        val orgX = orgPt.x.toFloat()
        val orgY = orgPt.y.toFloat()
        val currentBoundary: FloatArray = getBoundary(rect)
        val rotateBoundary = IntArray(8)

        rotateBoundary[4] = (orgX - (currentBoundary[0] - orgX)).toInt()
        rotateBoundary[5] = (orgY - (currentBoundary[1] - orgY)).toInt()
        rotateBoundary[6] = (orgX - (currentBoundary[2] - orgX)).toInt()
        rotateBoundary[7] = (orgY - (currentBoundary[3] - orgY)).toInt()
        rotateBoundary[0] = (orgX - (currentBoundary[4] - orgX)).toInt()
        rotateBoundary[1] = (orgY - (currentBoundary[5] - orgY)).toInt()
        rotateBoundary[2] = (orgX - (currentBoundary[6] - orgX)).toInt()
        rotateBoundary[3] = (orgY - (currentBoundary[7] - orgY)).toInt()

        return Rect(rotateBoundary[0], rotateBoundary[1], rotateBoundary[2], rotateBoundary[5])
    }

    fun getOrientationDisplayOffset(context: Context, nSensorOrientation: Int): Int {
        val display = (context.getSystemService(Context.WINDOW_SERVICE) as WindowManager).defaultDisplay
        val mDisplayOffset: Int = when (display.rotation) {
            Surface.ROTATION_0 -> 0
            Surface.ROTATION_90 -> 90
            Surface.ROTATION_180 -> 180
            Surface.ROTATION_270 -> 270
            else -> 0
        }
        return (nSensorOrientation - mDisplayOffset + 360) % 360
    }

    fun convertViewRegionToVideoFrameRegion(viewRegion: Rect, frameSize: Rect, nOrientationDisplayOffset: Int, szCameraView: Size): Rect {
        val convertRegion: Rect = when (nOrientationDisplayOffset) {
            90 -> boundaryRotate(Point(szCameraView.width / 2, szCameraView.height / 2), viewRegion, true)
            180 -> boundaryRotate180(Point(szCameraView.width / 2, szCameraView.height / 2), viewRegion)
            270 -> boundaryRotate(Point(szCameraView.width / 2, szCameraView.height / 2), viewRegion, false)
            else -> viewRegion
        }

        val nViewW = szCameraView.width
        val nViewH = szCameraView.height
        val fScaleH: Float
        val fScaleW: Float
        if (nOrientationDisplayOffset % 180 == 0) {
            fScaleH = 1.0f * frameSize.height() / nViewH
            fScaleW =  1.0f * frameSize.width() / nViewW
        } else {
            fScaleH = 1.0f * frameSize.height() / nViewW
            fScaleW = 1.0f * frameSize.width() / nViewH
        }

        val fScale = if (fScaleH > fScaleW) fScaleW else fScaleH
        val boxLeft = (convertRegion.left * fScale).toInt()
        val boxTop = (convertRegion.top * fScale).toInt()
        val boxWidth = (convertRegion.width() * fScale).toInt()
        val boxHeight = (convertRegion.height() * fScale).toInt()
        return Rect(boxLeft, boxTop, boxWidth + boxLeft, boxTop + boxHeight)
    }

    fun convertFrameRegionToViewRegion(frameRegion: Rect, frameSize: Rect, nOrientationDisplayOffset: Int, szCameraView: Size): Rect {
        var roateRect = frameRegion
        if (nOrientationDisplayOffset == 90) {
            roateRect = boundaryRotate(Point(frameSize.width() / 2, frameSize.height() / 2), frameRegion, false)
        } else if (nOrientationDisplayOffset == 180) {
            roateRect = boundaryRotate180(Point(frameSize.width() / 2, frameSize.height() / 2), frameRegion)
        } else if (nOrientationDisplayOffset == 270) {
            roateRect = boundaryRotate(Point(frameSize.width() / 2, frameSize.height() / 2), frameRegion, true)
        }
        val nViewW = szCameraView.width
        val nViewH = szCameraView.height
        val fScaleH = if (nOrientationDisplayOffset % 180 == 0) 1.0f * nViewH / frameSize.height() else 1.0f * nViewH / frameSize.width()
        val fScaleW = if (nOrientationDisplayOffset % 180 == 0) 1.0f * nViewW / frameSize.width() else 1.0f * nViewW / frameSize.height()
        val fScale = if (fScaleH > fScaleW) fScaleW else fScaleH
        val boxLeft = (roateRect.left * fScale).toInt()
        val boxTop = (roateRect.top * fScale).toInt()
        val boxWidth = (roateRect.width() * fScale).toInt()
        val boxHeight = (roateRect.height() * fScale).toInt()
        return Rect(boxLeft, boxTop, boxWidth + boxLeft, boxTop + boxHeight)
    }
}